plugins/woocommerce-admin/client/products/fields/category-field/category-field-item.tsx

```diff
@@ -0,0 +1,102 @@
+/**
+ * External dependencies
+ */
+import { CheckboxControl, Icon } from '@wordpress/components';
+import { useEffect, useState } from '@wordpress/element';
+import { chevronDown, chevronUp } from '@wordpress/icons';
+import { ProductCategory } from '@woocommerce/data';
+import classNames from 'classnames';
+
+export type CategoryTreeItem = {
+	data: ProductCategory;
+	children: CategoryTreeItem[];
+	parentID: number;
+	isOpen: boolean;
+};
+
+type CategoryFieldItemProps = {
+	item: CategoryTreeItem;
+	selectedIds: number[];
+	onSelect: ( item: ProductCategory ) => void;
+	items: Pick< ProductCategory, 'id' | 'name' >[];
+	highlightedIndex: number;
+	openParent?: () => void;
+};
+
+export const CategoryFieldItem: React.FC< CategoryFieldItemProps > = ( {
+	item,
+	selectedIds = [],
+	onSelect,
+	items,
+	highlightedIndex,
+	openParent,
+} ) => {
+	const [ isOpen, setIsOpen ] = useState( item.isOpen || false );
+	const index = items.findIndex( ( i ) => i.id === item.data.id );
+	const children = item.children.filter( ( child ) =>
+		items.includes( child.data )
+	);
+
+	useEffect( () => {
+		if ( highlightedIndex === index && children.length > 0 && ! isOpen ) {
+			setIsOpen( true );
+		} else if ( highlightedIndex === index && openParent ) {
+			// Make sure the parent is also open when the item is highlighted.
+			openParent();
+		}
+	}, [ highlightedIndex ] );
+
+	useEffect( () => {
+		if ( item.isOpen !== isOpen ) {
+			setIsOpen( item.isOpen );
+		}
+	}, [ item.isOpen ] );
+
+	return (
+		<div
```

Comment:
Thanks, I added the `ul` and `li` in. The `Menu` component uses the `ul` tag.
Also I like the above format, but it actually made things more complicated when trying to follow it, in particular with the managing of the opening and closing of each tree and such. So I kept the use of the `CategoryItem` and have that handle the children still.
I also updated that component to use the `ul` and `li` tags.

Reply:
Yes it's not an easy implementation with all the recursion logic. But I rather prefer containing all recursive children inside the same component for maintainability, reusability and readability. Even when I know this was made specific for categories it could be a great reusable component with nesting children to use in other place. Which means almost an entire refactor here. So lets keep things like they are, we'll have the time to talk about this again.